### PR TITLE
[READY]Refactors projectile trajectories, fixes piercing projectile inaccuracy and beam rifle visual deflections

### LINKED ISSUE #29315: Projectiles forcedodging through objects don't properly account for their new position and can drift off course
This is a very annoying bug/issue.

### LINKED ISSUE #32343: Beam rifles spaz the hell out when aiming at titanium walls due to their reflective nature.
This, combined with the totally unecessary "stepper vs instant" modes as well as being able to toggle between no zoom, locked directional, free directional, and center modes makes this gun feel extremely messy. Just have it be on instant center mode and have that be the only option, remove the action button abilities that only serve to make using this thing completely nauseating.

@kevinz000

### PR DESCRIPTION
**Pull request by silicons**
Fixes #32343
Fixes #29315
Projectiles now use a new /datum/point thingy, which measures in pixels from the edge of the map.
projectiles themselves use /datum/point/vector, designed for pixel-accurate angular trajectories.
Changes some loc = to forceMove()s.
:cl:
fix: Projectiles that pierce objects will no longer be thrown off course when doing so!
code: Projectiles now use /datum/point/vectors for trajectory calculations.
fix: Beam rifle reflections will no longer glitch out.
/:cl:

### CODE REVIEW

**Comment by optimumtact** [comment]
]wip[

File: code/datums/position_point_vector.dm [hunk 1]
```diff
@@ -0,0 +1,195 @@
+//Designed for things that need precision trajectories like projectiles.
+//Don't use this for anything that you don't absolutely have to use this with (like projectiles!) because it isn't worth using a datum unless you need accuracy down to decimal places in pixels.
+
+/proc/return_precise_position(atom/A)
```
Comment (RemieRichards) [inline code/datums/position_point_vector.dm]: You claim this is a precise position, yet you don't include matrices offsets.
0/10 go back to school mehki,

**Comment by ShizCalev** [comment]
Conflict.

File: code/datums/position_point_vector.dm [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+		x += maxx
+		move_zx -= 1
+	if(y < 0)
+		y += maxy
+		move_zy -= 1
+	if(x > maxx)
+		x -= maxx
+		move_zx += 1
+	if(y > maxy)
+		y += maxy
```
Comment (AnturK) [inline code/datums/position_point_vector.dm]: Shouldn't this be -=

File: code/datums/position_point_vector.dm [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+		_x = A.x
+		_y = A.y
+		_z = A.z
+		_pixel_x = A.pixel_x
+		_pixel_y = A.pixel_y
+	initialize_location(_x, _y, _z, _pixel_x, _pixel_y)
+
+/datum/point/proc/initialize_location(tile_x, tile_y, tile_z, p_x = 0, p_y = 0)
+	if(!isnull(tile_x))
+		x = ((tile_x - 1) * world.icon_size) + 16 + p_x
```
Comment (AnturK) [inline code/datums/position_point_vector.dm]: Using world.icon_size everywhere and straight 16 here is inconsistent.

**Review by AnturK — CHANGES_REQUESTED** [review]
Testmerge.

File: code/datums/position_point_vector.dm (hunk 2 above)
Comment (kevinz000) [reply code/datums/position_point_vector.dm]: kek

**Comment by ShiggyDiggyDo** [comment]
This broke emitter reflectors, I think. The emitter lasers won't change directions, just their sprite changes direction but not the laser.

**Comment by kevinz000** [comment]
fixed my shit.

**Comment by optimumtact** [comment]
one more test merge and this is probably safe to merge

File: code/datums/position_point_vector.dm [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+	var/x = 0
+	var/y = 0
+	var/z = 0
+	var/pixel_x = 0
+	var/pixel_y = 0
+
+/datum/position/proc/valid()
+	return x && y && z && !isnull(pixel_x) && !isnull(pixel_y)
+
+/datum/position/New(_x = 0, _y = 0, _z = 0, _pixel_x = 0, _pixel_y = 0)	//first argument can also be a /datum/point.
```
Comment (RemieRichards) [inline code/datums/position_point_vector.dm]: you're passing an atom above, but have no case here to handle an atom.

File: code/datums/position_point_vector.dm (hunk 1 above)
Comment (RemieRichards) [reply code/datums/position_point_vector.dm]: >months later
>still not done
>procs literally already exist to get x/y offsets from matrices

File: code/datums/position_point_vector.dm (hunk 4 above)
Comment (kevinz000) [reply code/datums/position_point_vector.dm]: oof.

File: code/datums/position_point_vector.dm (hunk 1 above)
Comment (kevinz000) [reply code/datums/position_point_vector.dm]: ;_;

**Comment by optimumtact** [comment]
stale reopen when ready or ping for reopen

File: code/datums/position_point_vector.dm (hunk 1 above)
Comment (optimumtact) [reply code/datums/position_point_vector.dm]: also make these into macros

Comment (optimumtact) [reply code/datums/position_point_vector.dm]: it's precise enough

File: code/datums/position_point_vector.dm (hunk 4 above)
Comment (optimumtact) [reply code/datums/position_point_vector.dm]: this was resolved for anyone following along

**Comment by Cyberboss** [comment]
Bagil tm

**Pull request merged**